### PR TITLE
[Bots] Add command support for raids, add additional arguments for actionables

### PR DESCRIPTION
**Pull request by nytmyr**
This adds raid support to actionables for bot commands.

This also addresses Issue #3567

If in a group, group based commands will function as normal, however if in a raid they will support raids and use the group of the bot that fits  the criteria.

This adds new actionables as well; namely, **ownerraid**, **byclass** and **byrace**.

**byclass** and **byrace** use the int of the chosen class/race. IE Shadowknight is 5, Barbarian is 2. **^create help** is an easy way for players to identify numbers associated with classes and races.

**targetgroup** will now select all bots that meet the criteria within a raid group in addition to the current functionality of groups.
**namesgroup** will now select all bots that meet the criteria within a raid group in addition to the current functionality of groups.
**ownerraid** will select all bots in the raid owned by the player.
**byclass** will selects all bots in the group or raid owned by the player that match the class.
**byrace** will selects all bots in the group or raid owned by the player that match the race

This also adds actionables to **^casterrange** and adds additional options to **^summon** and **^camp**

### CODE REVIEW

**Comment by nytmyr** [comment]
Added actionables for ^casterrange, added the new actionables to ^summon and corrected the mask for ^taunt to honor on/off toggles.

**Comment by Akkadius** [comment]
@Aeadoin @Kinglykrab can you guys take a look at this one when you can please

**Comment by Kinglykrab** [comment]
Needs to be rebased.

**Comment by nytmyr** [comment]
> Needs to be rebased.

Done...I think? 😂

**Review by Kinglykrab — APPROVED** [review]

**Comment by Kinglykrab** [comment]
Looks good to me, could probably do with some formatting around implicit statements and the like though.

File: zone/bot_command.cpp [hunk 1]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
 		if (!bot_owner)
 			return;
 
-		if (!bot_owner->GetGroup())
+		if (!bot_owner->GetGroup() && !bot_owner->GetRaid())
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1746]: Explicit scoping.
```cpp
if () {
}```

File: zone/bot_command.cpp [hunk 2]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
 			return;
 
-		std::list<Bot*> group_list;
-		bot_owner->GetGroup()->GetBotList(group_list);
-		for (auto member_iter : group_list) {
-			if (IsMyBot(bot_owner, member_iter))
-				sbl.push_back(member_iter);
+		if (bot_owner->IsRaidGrouped()) {
+			Raid* raid = bot_owner->GetRaid();
+			if (!raid)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1751]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;```

File: zone/bot_command.cpp [hunk 3]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
+				return;
+			uint32 raid_group = raid->GetGroup(bot_owner);
+			if (raid_group >= MAX_RAID_GROUPS) {
+				return;
+			}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1756]: Newline gap.

File: zone/bot_command.cpp [hunk 4]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
+			for (int x = 0; x < MAX_RAID_MEMBERS; x++) {
+				if (raid->members[x].member && entity_list.IsMobInZone(raid->members[x].member) && raid->members[x].group_number == raid_group) {
+					if (!MyBots::IsMyBot(bot_owner, raid->members[x].member->CastToBot()))
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1759]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;
```

File: zone/bot_command.cpp [hunk 5]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
-				sbl.push_back(member_iter);
+		if (bot_owner->IsRaidGrouped()) {
+			Raid* raid = bot_owner->GetRaid();
+			if (!raid)
+				return;
+			uint32 raid_group = raid->GetGroup(bot_owner);
+			if (raid_group >= MAX_RAID_GROUPS) {
+				return;
+			}
+			for (int x = 0; x < MAX_RAID_MEMBERS; x++) {
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1757]: Can probably do ```cpp
for (const auto& m : raid->members) {
    if (m.member) {
    ```

Don't think we have to check if it's in zone if we have a valid pointer.

File: zone/bot_command.cpp [hunk 6]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
+						continue;
+					sbl.push_back(raid->members[x].member->CastToBot());
+				}
+			}
+		}
+		else {
+			std::list<Bot*> group_list;
+			bot_owner->GetGroup()->GetBotList(group_list);
+			for (auto member_iter : group_list) {
+				if (IsMyBot(bot_owner, member_iter))
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1769]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 7]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
+					sbl.push_back(member_iter);
+			}
+		}
+
+		if (!clear_list)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1774]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 8]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
+			UniquifySBL(sbl);
+	}
+
+	static void PopulateSBL_ByMyRaidBots(Client* bot_owner, std::list<Bot*>& sbl, bool clear_list = true) {
+		if (clear_list)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1779]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;
```

File: zone/bot_command.cpp [hunk 9]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
+			sbl.clear();
+		if (!bot_owner)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1781]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 10]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
+			return;
+
+		if (!bot_owner->GetRaid())
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1784]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 11]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
+			return;
+
+		Raid* raid = bot_owner->GetRaid();
+		if (!raid)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1788]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 12]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
+			return;
+
+		for (int x = 0; x < MAX_RAID_MEMBERS; x++) {
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1791]: Can probably do
```cpp
for (const auto& m : raid->members) {
    if (m.member) {
```

Don't think we have to check if mob is in zone if we have a valid pointer.

File: zone/bot_command.cpp [hunk 13]
```diff
@@ -1743,14 +1743,57 @@ namespace MyBots
...
+			if (raid->members[x].member && entity_list.IsMobInZone(raid->members[x].member)) {
+				if (!MyBots::IsMyBot(bot_owner, raid->members[x].member->CastToBot()))
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1793]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;
```

File: zone/bot_command.cpp [hunk 14]
```diff
@@ -1764,14 +1807,34 @@ namespace MyBots
 			return;
 
 		auto target_mob = bot_owner->GetTarget();
-		if (!target_mob || !target_mob->GetGroup() || (!target_mob->IsClient() && !target_mob->IsBot()))
+		if (!target_mob || (!target_mob->GetGroup() && !target_mob->GetRaid()) || (!target_mob->IsClient() && !target_mob->IsBot()))
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1810]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 15]
```diff
@@ -1764,14 +1807,34 @@ namespace MyBots
...
-		for (auto member_iter : group_list) {
-			if (IsMyBot(bot_owner, member_iter))
-				sbl.push_back(member_iter);
+		if (bot_owner->IsRaidGrouped()) {
+			Raid* raid = bot_owner->GetRaid();
+			if (!raid)
+				return;
+			if (MyBots::IsMyBot(bot_owner, target_mob)) {
+				uint32 raid_group = raid->GetGroup(target_mob->CastToClient());
+				if (raid_group >= MAX_RAID_GROUPS) {
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1819]: Newline gap.

File: zone/bot_command.cpp [hunk 16]
```diff
@@ -1764,14 +1807,34 @@ namespace MyBots
...
 			return;
 
-		std::list<Bot*> group_list;
-		target_mob->GetGroup()->GetBotList(group_list);
-		for (auto member_iter : group_list) {
-			if (IsMyBot(bot_owner, member_iter))
-				sbl.push_back(member_iter);
+		if (bot_owner->IsRaidGrouped()) {
+			Raid* raid = bot_owner->GetRaid();
+			if (!raid)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1815]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;
```

File: zone/bot_command.cpp [hunk 17]
```diff
@@ -1764,14 +1807,34 @@ namespace MyBots
...
+						sbl.push_back(raid->members[x].member->CastToBot());
+					}
+				}
+			}
+		}
+		else {
+			std::list<Bot*> group_list;
+			bot_owner->GetGroup()->GetBotList(group_list);
+			for (auto member_iter : group_list) {
+				if (IsMyBot(bot_owner, member_iter))
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1835]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 18]
```diff
@@ -1793,14 +1856,32 @@ namespace MyBots
 				break;
 			}
 		}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1858]: Newline gap.

File: zone/bot_command.cpp [hunk 19]
```diff
@@ -1793,14 +1856,32 @@ namespace MyBots
...
-		if (!named_mob || !named_mob->GetGroup() || (!named_mob->IsClient() && !named_mob->IsBot()))
+		if (!named_mob || !MyBots::IsMyBot(bot_owner, named_mob) || (!named_mob->GetGroup() && !named_mob->GetRaid()) || (!named_mob->IsClient() && !named_mob->IsBot()))
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1859]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 20]
```diff
@@ -1793,14 +1856,32 @@ namespace MyBots
...
 			return;
 
-		std::list<Bot*> group_list;
-		named_mob->GetGroup()->GetBotList(group_list);
-		for (auto member_iter : group_list) {
-			if (IsMyBot(bot_owner, member_iter))
-				sbl.push_back(member_iter);
+		if (bot_owner->IsRaidGrouped()) {
+			Raid* raid = bot_owner->GetRaid();
+			if (!raid)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1864]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;
```

File: zone/bot_command.cpp [hunk 21]
```diff
@@ -1793,14 +1856,32 @@ namespace MyBots
...
+				return;
+			uint32 raid_group = raid->GetGroup(named_mob->CastToClient());
+			if (raid_group >= MAX_RAID_GROUPS) {
+				return;
+			}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1869]: Newline gap.

File: zone/bot_command.cpp [hunk 22]
```diff
@@ -1793,14 +1856,32 @@ namespace MyBots
...
+			for (int x = 0; x < MAX_RAID_MEMBERS; x++) {
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1870]: Can probably do
```cpp
for (const auto& m : raid->members) {
    if (m.member) {
```

Don't think we have to check if mob is in zone if we have a valid pointer.

File: zone/bot_command.cpp [hunk 23]
```diff
@@ -1793,14 +1856,32 @@ namespace MyBots
...
+						continue;
+					sbl.push_back(raid->members[x].member->CastToBot());
+				}
+			}
+		}
+		else {
+			std::list<Bot*> group_list;
+			bot_owner->GetGroup()->GetBotList(group_list);
+			for (auto member_iter : group_list) {
+				if (IsMyBot(bot_owner, member_iter))
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1882]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 24]
```diff
@@ -1896,7 +1977,40 @@ namespace MyBots
 		sbl.remove(nullptr);
 	}
 
-}
+	static void PopulateSBL_BySpawnedBotsClass(Client * bot_owner, std::list<Bot*> &sbl, uint16 cls, bool clear_list = true) {
+		if (clear_list)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1981]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;
```

File: zone/bot_command.cpp [hunk 25]
```diff
@@ -1896,7 +1977,40 @@ namespace MyBots
...
+			sbl.clear();
+		if (!bot_owner || !cls)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1983]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 26]
```diff
@@ -1896,7 +1977,40 @@ namespace MyBots
...
+			return;
+
+		auto selectable_bot_list = entity_list.GetBotsByBotOwnerCharacterID(bot_owner->CharacterID());
+		for (auto bot_iter : selectable_bot_list) {
+			if (bot_iter->GetClass() != cls) {
+				continue;
+			}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1990]: Newline gap.

File: zone/bot_command.cpp [hunk 27]
```diff
@@ -1896,7 +1977,40 @@ namespace MyBots
...
+			sbl.push_back(bot_iter);
+		}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1992]: Newline gap.

File: zone/bot_command.cpp [hunk 28]
```diff
@@ -1896,7 +1977,40 @@ namespace MyBots
...
+		if (!clear_list)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1993]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 29]
```diff
@@ -1896,7 +1977,40 @@ namespace MyBots
...
+			UniquifySBL(sbl);
+	}
+
+	static void PopulateSBL_BySpawnedBotsRace(Client* bot_owner, std::list<Bot*>& sbl, uint16 race, bool clear_list = true) {
+		if (clear_list)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:1998]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;
```

File: zone/bot_command.cpp [hunk 30]
```diff
@@ -1896,7 +1977,40 @@ namespace MyBots
...
+			sbl.clear();
+		if (!bot_owner || !race)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:2000]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 31]
```diff
@@ -1896,7 +1977,40 @@ namespace MyBots
...
+			return;
+
+		auto selectable_bot_list = entity_list.GetBotsByBotOwnerCharacterID(bot_owner->CharacterID());
+		for (auto bot_iter : selectable_bot_list) {
+			if (bot_iter->GetBaseRace() != race) {
+				continue;
+			}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:2007]: Newline gap.

File: zone/bot_command.cpp [hunk 32]
```diff
@@ -1896,7 +1977,40 @@ namespace MyBots
...
+			sbl.push_back(bot_iter);
+		}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:2009]: Newline gap.

File: zone/bot_command.cpp [hunk 33]
```diff
@@ -1896,7 +1977,40 @@ namespace MyBots
...
+		if (!clear_list)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:2010]: Explicit scoping.
```cpp
if () {
}
```

File: zone/bot_command.cpp [hunk 34]
```diff
@@ -2216,24 +2362,47 @@ namespace ActionableBots
...
-			return static_cast<Bot*>(member_iter);
+		if (bot_owner->IsRaidGrouped()) {
+			Raid* raid = bot_grouped_player->GetRaid();
+			if (!raid)
+				return nullptr;
+			uint32 raid_group = raid->GetGroup(bot_grouped_player);
+			if (raid_group >= MAX_RAID_GROUPS) {
+				return nullptr;
+			}
+			for (int x = 0; x < MAX_RAID_MEMBERS; x++) {
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:2380]: Can probably do
```cpp
for (const auto& m : raid->members) {
    if (m.member) {
```

Don't think we have to check if mob is in zone if we have a valid pointer.

File: zone/bot_command.cpp [hunk 35]
```diff
@@ -2216,24 +2362,47 @@ namespace ActionableBots
...
 
-			return static_cast<Bot*>(member_iter);
+		if (bot_owner->IsRaidGrouped()) {
+			Raid* raid = bot_grouped_player->GetRaid();
+			if (!raid)
+				return nullptr;
+			uint32 raid_group = raid->GetGroup(bot_grouped_player);
+			if (raid_group >= MAX_RAID_GROUPS) {
+				return nullptr;
+			}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:2379]: Newline gap.

File: zone/bot_command.cpp [hunk 36]
```diff
@@ -2216,24 +2362,47 @@ namespace ActionableBots
...
-				continue;
-			if (member_iter->GetClass() != cls)
-				continue;
-			if (petless && member_iter->GetPet())
-				continue;
 
-			return static_cast<Bot*>(member_iter);
+		if (bot_owner->IsRaidGrouped()) {
+			Raid* raid = bot_grouped_player->GetRaid();
+			if (!raid)
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:2374]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;
```

File: zone/bot_command.cpp [hunk 37]
```diff
@@ -2216,24 +2362,47 @@ namespace ActionableBots
...
+				return nullptr;
+			}
+			for (int x = 0; x < MAX_RAID_MEMBERS; x++) {
+				if (raid->members[x].member && entity_list.IsMobInZone(raid->members[x].member) && raid->members[x].group_number == raid_group) {
+					if (!MyBots::IsMyBot(bot_owner, raid->members[x].member->CastToBot()))
+						continue;
+					if (raid->members[x].member->GetClass() != cls)
+						continue;
+					if (petless && raid->members[x].member->GetPet())
+						continue;
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:2387]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;
```

File: zone/bot_command.cpp [hunk 38]
```diff
@@ -2216,24 +2362,47 @@ namespace ActionableBots
...
+		else {
+			std::list<Mob*> group_list;
+			bot_grouped_player->GetGroup()->GetMemberList(group_list);
+			for (auto member_iter : group_list) {
+				if (!MyBots::IsMyBot(bot_owner, member_iter))
+					continue;
+				if (member_iter->GetClass() != cls)
+					continue;
+				if (petless && member_iter->GetPet())
+					continue;
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:2402]: Explicit scoping and newline gap.
```cpp
if () {
}

sadasda;
```

File: zone/bot_command.cpp [hunk 39]
```diff
@@ -9955,49 +10184,107 @@ void bot_command_caster_range(Client* c, const Seperator* sep)
...
-	uint32 crange = 0;
-	if (sep->IsNumber(1)) {
-		crange = atoi(sep->arg[1]);
-		if (crange >= 0 && crange <= 300) {
+	Bot* first_found = nullptr;
+	int success_count = 0;
+	for (auto my_bot : sbl) {
+		if (!IsCasterClass(my_bot->GetClass()) && !IsHybridClass(my_bot->GetClass())) {
+			continue;
+		}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:10237]: Newline gap.

File: zone/bot_command.cpp [hunk 40]
```diff
@@ -9955,49 +10184,107 @@ void bot_command_caster_range(Client* c, const Seperator* sep)
...
+		if (!first_found) {
+			first_found = my_bot;
+		}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:10240]: Newline gap.

File: zone/bot_command.cpp [hunk 41]
```diff
@@ -9955,49 +10184,107 @@ void bot_command_caster_range(Client* c, const Seperator* sep)
...
+					Chat::White,
+					fmt::format(
+						"{} for '{}'",
+						BotDatabase::fail::SaveBotCasterRange(),
+						my_bot->GetCleanName()
+					).c_str()
+				);
 			}
 		}
+	}
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:10265]: Newline gap.

**Comment by Kinglykrab** [comment]
Minor comments, other than that it functions well.

**Pull request merged**